pytential/symbolic/pde/reduce_fmms.py

```diff
@@ -0,0 +1,320 @@
+__copyright__ = "Copyright (C) 2021 Isuru Fernando"
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+from sumpy.kernel import (AxisTargetDerivative, AxisSourceDerivative)
+
+from pymbolic.mapper import Mapper
+from pymbolic.primitives import Product
+from pymbolic.interop.sympy import PymbolicToSympyMapper, SympyToPymbolicMapper
+import sympy
+
+from collections import defaultdict
+
+
+__all__ = (
+    "reduce_number_of_fmms",
+    )
+
+
+def reduce_number_of_fmms(int_gs, source_dependent_variables):
+    """
+    Reduce the number of FMMs needed for a system of expressions with
+    :class:`~pytential.symbolic.primitives.IntG` objects.
+
+    This is done by converting the `IntG` expression to a matrix of polynomials
+    with d variables corresponding to d dimensions and each polynomial represents
+    a derivative operator. All the properties of derivative operator that we want
+    are reflected in the properties of the polynomial including addition,
+    multiplication and exact polynomial division.
+
+    This matrix is factored into two matrices where the left hand side matrix
+    represents a transformation at target and the right hand side matrix represents
+    a transformation at source.
+
+    If the expressions given are not linear, then the input expressions are
+    returned as is.
+
+    :arg source_dependent_variables: When reducing FMMs, consider only these
+        variables as dependent on source.
```

Comment:
* Give examples.
* State that target-dependent variables are assumed to be not present.

Reply:
Done in c4068e9